libraries/rush-lib/src/logic/PackageJsonUpdater.ts

```diff
@@ -151,6 +181,196 @@ export class PackageJsonUpdater {
     this._terminal = new Terminal(this._terminalProvider);
   }
 
+  /**
+   * Upgrade dependencies to a particular project, or across specified projects. This is the core business logic for
+   * "rush upgrade-interactive".
+   */
+  public async doRushUpgradeAsync(options: IPackageJsonUpdaterRushUpgradeOptions): Promise<void> {
+    const { projects, packagesToAdd, updateOtherPackages, skipUpdate, debugInstall, variant } = options;
+    const { DependencyAnalyzer } = await import('./DependencyAnalyzer');
+    const dependencyAnalyzer: DependencyAnalyzer = DependencyAnalyzer.forRushConfiguration(
+      this._rushConfiguration
+    );
+    const {
+      allVersionsByPackageName,
+      implicitlyPreferredVersionByPackageName,
+      commonVersionsConfiguration
+    }: IDependencyAnalysis = dependencyAnalyzer.getAnalysis(variant);
+
+    const dependenciesToUpdate: Record<string, string> = {};
+    const devDependenciesToUpdate: Record<string, string> = {};
+
+    for (const { moduleName, latest: latestVersion, packageJson, devDependency } of packagesToAdd) {
+      const inferredRangeStyle: SemVerStyle = this._cheaplyDetectSemVerRangeStyle(packageJson);
+      const implicitlyPreferredVersion: string | undefined =
+        implicitlyPreferredVersionByPackageName.get(moduleName);
+
+      const explicitlyPreferredVersion: string | undefined =
+        commonVersionsConfiguration.preferredVersions.get(moduleName);
+
+      const version: string = await this._getNormalizedVersionSpec(
+        projects,
+        moduleName,
+        latestVersion,
+        implicitlyPreferredVersion,
+        explicitlyPreferredVersion,
+        inferredRangeStyle
+      );
+
+      if (devDependency) {
+        devDependenciesToUpdate[moduleName] = version;
+      } else {
+        dependenciesToUpdate[moduleName] = version;
+      }
```

Comment:
Whether a given external package should be a `devDependency` or a `dependency` really depends on the nature of the consuming project. This logic is outside of the `projects` loop, which suggests that you are setting the type globally.